vortex-serde/src/layouts/read/filtering.rs

```diff
@@ -21,29 +21,31 @@ pub struct RowFilter {
 
 impl RowFilter {
     pub fn new(expr: Arc<dyn VortexExpr>) -> Self {
-        let conjunction = split_conjunction(&expr)
-            .into_iter()
-            .filter(expr_is_filter)
-            .collect();
-
+        let conjunction = split_conjunction(&expr);
         Self { conjunction }
     }
 
     /// Evaluate the underlying filter against a target array, returning a boolean mask
     pub fn evaluate(&self, target: &Array) -> VortexResult<Array> {
-        let mut mask = BoolArray::from(vec![true; target.len()]).into_array();
-        for expr in self.conjunction.iter() {
-            let new_mask = expr.evaluate(target)?;
-            mask = and(new_mask, mask)?;
-
+        let mut filter_iter = self.conjunction.iter();
+        let mut mask = filter_iter
+            .next()
+            .vortex_expect("must have at least one predicate")
+            .evaluate(target)?;
+        for expr in filter_iter {
             if mask.statistics().compute_true_count().unwrap_or_default() == 0 {
-                return Ok(
-                    BoolArray::from_vec(vec![false; target.len()], Validity::AllValid).into_array(),
-                );
+                return BoolArray::try_new(
+                    BooleanBuffer::new_unset(target.len()),
+                    Validity::AllInvalid,
+                )
+                .map(IntoArray::into_array);
             }
+
+            let new_mask = expr.evaluate(target)?;
```

Comment:
since we're evaluating sequentially, it might be worth it a FLUP to do the selection pushdown thing and filter the target before evaluating the next mask. def more complicated though with scatter/gather

Reply:
This is one of the followups to the pushdown, we need a filter with a mask (or list of masks)